Add async processing support to support async plugins

mikesir87 opened the pull request:
Resolves #680

<!--
  Please check the needed checkboxes ([ ] -> [x]). Leave the
  comments as they are, they won’t show on GitHub.
  We are excited about pull requests, but please try to limit the scope, provide
  a general description of the changes, and remember, it’s up to you to convince
  us to land it.
-->

### Initial checklist

*   [x] I read the support docs <!-- https://github.com/remarkjs/.github/blob/main/support.md -->
*   [x] I read the contributing guide <!-- https://github.com/remarkjs/.github/blob/main/contributing.md -->
*   [x] I agree to follow the code of conduct <!-- https://github.com/remarkjs/.github/blob/main/code-of-conduct.md -->
*   [x] I searched issues and couldn’t find anything (or linked relevant results below) <!-- https://github.com/search?q=user%3Aremarkjs&type=Issues -->
*   [x] If applicable, I’ve added docs and tests

### Description of changes

Added an `async` option to the component. If true, runs the renderer using `processor.run` instead of `processor.runSync` (which causes an error be thrown).

<!--do not edit: pr-->

mikesir87 commented:
Docs updated and added a simple async plugin to validate it both executes and the overall rendering still works.

adamelmore commented:
Is this PR still alive? I'm also looking for async plugin support and wondering if this effort was ongoing. Happy to help if needed.

wooorm commented:
There are still some more todos from https://github.com/remarkjs/react-markdown/pull/682#issuecomment-1090068072 as I understand it.

mikesir87 commented:
As far as I'm aware (it's been a little while though), I addressed all of the concerns in the comment. 

However, I did find one additional bug... if the markdown changes (eg, you have a live editor or a component is reused with new markdown), the new markdown isn't rendered. That's the challenge of trying to use the same component for both async and synchronous work. One idea is to make a completely separate component that renders asynchronously, allowing me to use `useState` and other hooks properly (I can't use those now because you can't have conditional hooks and using them would make all rendering async).

wooorm commented:
> This is especially needed

Why?

> as highlighted here

What you link is Next.js-specific documentation, about their own API for pages and layouts, which indeed can be async, I think that was already possible, and I don’t see how that relates to React?

reypm commented:
hey folks, what is the status of this PR will this be merged any time soon?

ChristianMurphy commented:
@reypm you can find the status by reading the comments directly above yours.
This is blocked by unresolved requests https://github.com/remarkjs/react-markdown/pull/682#issuecomment-1133627232 and by unresolved discussion https://github.com/remarkjs/react-markdown/pull/682#issuecomment-1334815897.
It will be merged when and if it is ready.

simonpfish commented:
Hey @wooorm @ChristianMurphy! How can I help land this PR?

simonpfish commented:
We want to use a couple of async plugins in cookbook.openai.com, and this is currently blocking us. Happy to help iterate on it to be able to merge.

simonpfish commented:
Got it! I ended up just using `unified` directly and `rehype-react`

lovettbarron commented:
Is there any progress or blockers on this one? I've encountered it a few times now: initially with a plugin I'm developing, and now again with [rehypejs/rehype-starry-night](https://github.com/rehypejs/rehype-starry-night)

wooorm commented:
see the earlier comments please :)

wooorm commented:
Investigated this for a bit last week, an `async` prop which makes it return `Promise<JSX.Element>`. This is still not supported normally in React 19. Plus, conditionally returning `Promise<JSX.Element> | JSX.Element` breaks all the types of React 🤷‍♂️ So I gave up.

x5engine commented:
This is how you kill a famous amazing package because you do a breaking change and dont want to accept background compability for now until you have another way to make old plugins work or updated!

Sad

muhammadsr commented:
@simonpfish saved the day for me. Thanks

wooorm commented:
Released in [`9.1.0`](https://github.com/remarkjs/react-markdown/releases/tag/9.1.0)!
